### PR TITLE
Trwalke/ios background thread fix

### PR DESCRIPTION
**Pull request by trwalke**
Resolved issue with iOS 11.3 where network resources are reclaimed by the system when sending an app to the background.

437419

https://identitydivision.visualstudio.com/DevEx/_workitems/edit/437419

**Repro steps**
1. Launch any ios app that uses adal.
2. Try to authenticate with an account that requires MFA.
3. Switch to the broker app for MFA when prompted.(if the broker app is on another device, simply send the app      to the background with the home button)
4. Switch back to the ios app
5. Authenticate with MFA 
6. Notice that the app hangs waiting for MFA to be completed

### CODE REVIEW

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 1]
```diff
@@ -36,6 +38,9 @@ internal class WebUI : IWebUI
         private static SemaphoreSlim returnedUriReady;
         private static AuthorizationResult authorizationResult;
         private PlatformParameters parameters;
+        private nint taskId = 0;
+        private static bool inBackground;
+        NSObject DidEnterBackgroundNotification, WillEnterForegroundNotification;
```
Comment (henrik-me) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: private?

Comment (trwalke) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: fixed

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 2]
```diff
@@ -44,6 +49,30 @@ public WebUI(IPlatformParameters parameters)
...
+        void OnMoveToBackground(NSNotification notification)
+        {
+            taskId = UIApplication.SharedApplication.BeginBackgroundTask(() => {
+                inBackground = false;
+                UIApplication.SharedApplication.EndBackgroundTask(taskId);
+            });
+
+            //After iOS 11.3, it is neccesary to keep a background task running while moving an app to the background in order to prevent the system from reclaiming network resources from the app. 
+            //This will prevent authentication from failing while the application is moved to the background while waiting for MFA to finish.
+            new Task(() => {
```
Comment (oldalton) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: what does the Task do? I see you're already doing BeginBackgroundTask, so why is this necessary?

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 3]
```diff
@@ -44,6 +49,30 @@ public WebUI(IPlatformParameters parameters)
...
+                inBackground = true;
+                while (inBackground) { }
```
Comment (oldalton) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: I'm not sure why this is necessary

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 4]
```diff
@@ -89,6 +118,9 @@ public void Authenticate(Uri authorizationUri, Uri redirectUri, CallState callSt
         private void CallbackMethod(AuthorizationResult result)
         {
             SetAuthorizationResult(result);
+            inBackground = false;
+            DidEnterBackgroundNotification.Dispose();
```
Comment (oldalton) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: will dispose also unregister from NotificationCenter?

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 5]
```diff
@@ -44,6 +49,30 @@ public WebUI(IPlatformParameters parameters)
             {
                 throw new ArgumentException("parameters should be of type PlatformParameters", "parameters");
             }
+
+            DidEnterBackgroundNotification = NSNotificationCenter.DefaultCenter.AddObserver(UIApplication.DidEnterBackgroundNotification, OnMoveToBackground);
+            WillEnterForegroundNotification = NSNotificationCenter.DefaultCenter.AddObserver(UIApplication.WillEnterForegroundNotification, OnMoveToForeground);
```
Comment (henrik-me) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: rest of the file uses this... what about here?

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 6]
```diff
@@ -44,6 +49,30 @@ public WebUI(IPlatformParameters parameters)
...
+        }
+
+        void OnMoveToBackground(NSNotification notification)
+        {
+            taskId = UIApplication.SharedApplication.BeginBackgroundTask(() => {
```
Comment (henrik-me) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: >taskId [](start = 12, length = 6)

this. ?

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs (hunk 4 above)
Comment (henrik-me) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: pls. see also OnMoveToForeground... shouldn't that also clean the background task and clear the notifications?

---
In reply to: [188468356](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1053#discussion_r188468356) [](ancestors = 188468356)

**Review by henrik-me — CHANGES_REQUESTED** [review]
:clock1:

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 7]
```diff
@@ -44,6 +49,30 @@ public WebUI(IPlatformParameters parameters)
...
+            new Task(() => {
+                inBackground = true;
+                while (inBackground) { }
+                UIApplication.SharedApplication.EndBackgroundTask(taskId);
+            }).Start();
+        }
+
+        void OnMoveToForeground(NSNotification notification)
+        {
+            inBackground = false;
```
Comment (oldalton) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: I'd rather see background task being ended here instead of having a while loop and relying on a boolean, unless there's a good reason to do differently

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 8]
```diff
@@ -44,6 +49,30 @@ public WebUI(IPlatformParameters parameters)
...
+
+            DidEnterBackgroundNotification = NSNotificationCenter.DefaultCenter.AddObserver(UIApplication.DidEnterBackgroundNotification, OnMoveToBackground);
+            WillEnterForegroundNotification = NSNotificationCenter.DefaultCenter.AddObserver(UIApplication.WillEnterForegroundNotification, OnMoveToForeground);
+        }
+
+        void OnMoveToBackground(NSNotification notification)
+        {
+            taskId = UIApplication.SharedApplication.BeginBackgroundTask(() => {
+                inBackground = false;
+                UIApplication.SharedApplication.EndBackgroundTask(taskId);
```
Comment (oldalton) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: I'd suggest setting taskId to UIBackgroundTaskInvalid if there's such thing in Xamarin

**Review by oldalton — COMMENTED** [review]
A couple of questions and suggestions

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (oldalton) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: In iOS we are clearing foreground notification here and we're only listening to foreground transition after app goes to background. We still need to listen for background transition at this point in case user backgrounds the app twice and clean it up when we're done with auth or the class is dealloced.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 9]
```diff
@@ -44,6 +49,30 @@ public WebUI(IPlatformParameters parameters)
...
             }
+
+            DidEnterBackgroundNotification = NSNotificationCenter.DefaultCenter.AddObserver(UIApplication.DidEnterBackgroundNotification, OnMoveToBackground);
+            WillEnterForegroundNotification = NSNotificationCenter.DefaultCenter.AddObserver(UIApplication.WillEnterForegroundNotification, OnMoveToForeground);
+        }
+
+        void OnMoveToBackground(NSNotification notification)
+        {
+            taskId = UIApplication.SharedApplication.BeginBackgroundTask(() => {
+                inBackground = false;
```
[the remark this answers is not in the document]
Comment (trwalke) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: This is an expiration handler for this task that is called if it expires due to timeout or some other reason. Thus am setting this to false to ensure the background task ends when this if handler is called. I will explain in detail tomorrow.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs (hunk 2 above)
Comment (trwalke) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: @oldalton The BeginBackgroundTask does not actually start the task. it simply notifies the system that the following code should not be terminated. it also registers a task id for it. The space in between BeginBackgroundTask and EndBackgroundTask is where the task is created and executed.

Comment (oldalton) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: Yes, I know. But what does that Task do? If I'm reading the code correctly, it just starts a while loop and waits for the boolean to change. So, if it doesn't do anything, why is it necessary?

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 10]
```diff
@@ -118,9 +112,15 @@ public void Authenticate(Uri authorizationUri, Uri redirectUri, CallState callSt
         private void CallbackMethod(AuthorizationResult result)
         {
             SetAuthorizationResult(result);
-            inBackground = false;
-            DidEnterBackgroundNotification.Dispose();
-            WillEnterForegroundNotification.Dispose();
```
Comment (trwalke) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: Moving notification registration to new dispose method that is called when the webUI is closed.

**Review by henrik-me — APPROVED** [review]
:shipit:

**Comment by henrik-me** [comment]
Please add a comment on manual tests completed to ensure it works as intended.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 11]
```diff
@@ -44,6 +48,25 @@ public WebUI(IPlatformParameters parameters)
...
+            this.willEnterForegroundNotification = NSNotificationCenter.DefaultCenter.AddObserver(UIApplication.WillEnterForegroundNotification, OnMoveToForeground);
+        }
+
+        void OnMoveToBackground(NSNotification notification)
+        {
+            //After iOS 11.3, it is neccesary to keep a background task running while moving an app to the background in order to prevent the system from reclaiming network resources from the app. 
+            //This will prevent authentication from failing while the application is moved to the background while waiting for MFA to finish.
+            this.taskId = UIApplication.SharedApplication.BeginBackgroundTask(() => {
+                if (this.taskId != UIApplication.BackgroundTaskInvalid)
+                    UIApplication.SharedApplication.EndBackgroundTask(this.taskId);
```
Comment (oldalton) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: this should set taskId to BackgroundTaskInvalid too

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 12]
```diff
@@ -44,6 +48,25 @@ public WebUI(IPlatformParameters parameters)
...
+            });
+        }
+
+        void OnMoveToForeground(NSNotification notification)
+        {
+            if (this.taskId != UIApplication.BackgroundTaskInvalid)
+                UIApplication.SharedApplication.EndBackgroundTask(this.taskId);
```
Comment (oldalton) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: this should set it invalid task too

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 13]
```diff
@@ -90,5 +113,14 @@ private void CallbackMethod(AuthorizationResult result)
         {
             SetAuthorizationResult(result);
         }
+
+        public void Dispose()
+        {
+            if (this.taskId != UIApplication.BackgroundTaskInvalid)
+                UIApplication.SharedApplication.EndBackgroundTask(this.taskId);
```
Comment (oldalton) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: why this is necessary?

**Review by oldalton — DISMISSED** [review]
I didn't see task being set to invalid task

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs (hunk 13 above)
Comment (trwalke) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: I just wanted to be certain that the events are unregistered if the webUI is disposed before the app is returned to the foreground somehow. probably not needed.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs (hunk 11 above)
Comment (trwalke) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: fixed

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs (hunk 12 above)
Comment (trwalke) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: fixed

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs [hunk 14]
```diff
@@ -90,5 +119,11 @@ private void CallbackMethod(AuthorizationResult result)
         {
             SetAuthorizationResult(result);
         }
+
+        public void Dispose()
+        {
+            this.didEnterBackgroundNotification.Dispose();
```
Comment (oldalton) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: just in case, can you test and make sure this is called? According to Apple: https://developer.apple.com/documentation/foundation/nsnotificationcenter/1415360-addobserver?language=objc

If your app targets iOS 9.0 and later or macOS 10.11 and later, you don't need to unregister an observer in its dealloc method. Otherwise, **you should call removeObserver:name:object: before observer or any object passed to this method is deallocated.**

Comment (trwalke) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/WebUI.cs]: Verified already. With our our library creating a new WebUI for each authentication call, I was able to verify that this dispose method is called when the user successfully authenticates, authentication fails and when the user cancels the authentication and closes the WebUI. This was tested on a device.

**Comment by trwalke** [comment]
@oldalton any other comments or am I ok to merge?

**Comment by trwalke** [comment]
Manual testing was done to ensue that the functionality works as expected. 
The reproduction steps outlined here https://identitydivision.visualstudio.com/DevEx/_workitems/edit/437419
were used for manual testing. 

1. Launch any ios app that uses adal.
2. Try to authenticate with an account that requires MFA.
3. Switch to the broker app for MFA when prompted.(if the broker app is on another device, simply send the app to the background with the home button)
4. Switch back to the ios app
5. Authenticate with MFA 
6. _**Notice that the app hangs waiting for MFA to be completed**_
The application no longer hangs on step 6

**Comment by oldalton** [comment]
@trwalke The only other little suggestion I have is to check for invalid task before starting the background task, otherwise it looks fine to me.
